Create IRoutingTopology with factory delegate

bording opened the pull request:
This adds a `UseRoutingTopology` overload that accepts a factory delegate in order to allow creating instances with a non-default constructor.

The original `UseRoutingTopology` method and `UseDirectRoutingTopology` have been reimplemented using the new factory approach. The original `UseRoutingTopology` method should be obsoleted on the next major release.

Since both `ConventionalRoutingTopology` and `DirectRoutingTopology` receive `useDurableExchanges` in their constructors, the factory delegate also passes this in. External implementations should now be at feature parity with our internal topologies.

The routing topology creation  in `RabbitMQTransportInfrastructure` has been updated and simplified as part of this as well.

This addresses the concern raised in https://github.com/Particular/NServiceBus.RabbitMQ/issues/248#issuecomment-254238054 by @volak.

bording commented:
@adamralph I made a change to the factory parameter description which attempts to explain what the parameter is for.

Also, if we decide we can go ahead and do it now, I'll add the obsolete to the old method.

adamralph commented:
@bording thanks, the docs look good.

It seems that we can indeed obsolete the old method now with a warning, so if you add that then I'll be 👍 to merge this.

bording commented:
I pushed up a couple more changes. First, I realized that the new method doesn't need to be generic, so I fixed that.

The obsolete attribute I added generates the following message:

'RabbitMQTransportSettingsExtensions.UseRoutingTopology<T>(TransportExtensions<RabbitMQTransport>) is obsolete: 'Use `RabbitMQTransportSettingsExtensions.UseRoutingTopology(TransportExtensions<RabbitMQTransport> transportExtensions, Func<bool, IRoutingTopology>)` instead. Will be treated as an error from version 5.0.0. Will be removed in version 5.0.0.'

adamralph commented:
@bording 👍 is it worth squashing the commits before merging?

bording commented:
@adamralph I have squashed everything into a single commit.

FYI, one minor tweak I did include as part of this squashed commit is I changed the `RemoveInVersion` value of the obsolete attribute to 6.0 instead of 5.0. After having some other conversations around obsoletes with the core maintainers, it seems more correct to me to go ahead and have this around in v5 to let us provide an error message, before removing it entirely in v6.

adamralph commented:
LGTM. Will merge when green.
